nginx: fix failure to serve on aliases with regex captures

yorickvP opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

It turns out the [previous patch](https://github.com/NixOS/nixpkgs/pull/48337) did not take in consideration that you could do `alias a/$1/b`. This means `realpath` would return `NULL` on this and nginx would return a 500 error instead.

###### Things done

This PR does the following:
- add a test for this case
- use the actual file path instead of the configured root dir, so symlinks to outside of the nix store work as expected
- do not throw out the whole http request if `realpath` returns `NULL`

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @edef1c @yegortimoshenko @aszlig @grahamc @domenkozar

yorickvP committed:
nixos/tests/nginx: add (failing) test for regex captured aliases

mmahut commented:
@GrahamcOfBorg test nginx

7c6f434c reviewed (approved):
But maybe use `git mv` on the patch — the patch gets complicated and it is hard to see the changes to the patch…

lukateras reviewed (commented):
Would be able to review ~after `git mv`~ when there is a line-by-line diff.

yorickvP commented:
Git mv is a shorthand for git add && git rm. It won't help rename
detection. I can split this patch into a change and a rename commit,
though.

On Tue, Aug 13, 2019, 10:47 Yegor Timoshenko <notifications@github.com>
wrote:

> *@yegortimoshenko* commented on this pull request.
>
> Would be able to review after git mv.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/66532?email_source=notifications&email_token=AAE57JG2N6UGVP3YMPRFQFTQEJYLDA5CNFSM4ILCCUGKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCBLQXIY#pullrequestreview-274140067>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAE57JBIDUHTJCUMLUJ2UALQEJYLDANCNFSM4ILCCUGA>
> .
>

yegortimoshenko commented:
It seems that my choice of file name and format was unfortunate, maybe this could be better off as a diff (instead of a patch) and without Nginx version in the file name?

aszlig commented on pkgs/servers/http/nginx/nix-etag-1.16.0.patch:
```diff
@@ -0,0 +1,163 @@
+ ngx_int_t ngx_http_set_content_type(ngx_http_request_t *r);
+ void ngx_http_set_exten(ngx_http_request_t *r);
+-ngx_int_t ngx_http_set_etag(ngx_http_request_t *r);
++ngx_int_t ngx_http_set_etag(ngx_http_request_t *r, ngx_str_t *path);
```
This is a bit unfortunate here, because this changes the API for third-party extension modules.

yegortimoshenko replied:
Same concern, I don't think we should ever change API of any programs that we patch (i.e. no headers changes).

aszlig commented on pkgs/servers/http/nginx/nix-etag-1.16.0.patch:
```diff
@@ -0,0 +1,163 @@
++
++        if (etag->value.data == NULL) {
++            etag->hash = 0;
++            return NGX_ERROR;
```
Memleak: `ngx_free(real);`

aszlig commented on pkgs/servers/http/nginx/nix-etag-1.16.0.patch:
```diff
@@ -0,0 +1,163 @@
++                          - etag->value.data;
++    }
+ 
+     r->headers_out.etag = etag;
```
Same here: `ngx_free(real);`

aszlig commented on pkgs/servers/http/nginx/nix-etag-1.16.0.patch:
```diff
@@ -0,0 +1,163 @@
++        }
++
++        ngx_memcpy(etag->value.data, ptr1, etag->value.len);
++        ngx_free(real);
```
This isn't needed if you just move it past the `else` branch below.

aszlig reviewed (changes requested):
As mentioned by @yegortimoshenko, it would probably make sense to just use a plain diff and either keep the versioned name (it still applies on version 1.15.4) or drop it entirely.

Apart from the mentioned memleaks, the API change is something I'd try to avoid as we don't know which external modules people might use.

yorickvP committed:
nginx: rename nix-etag-1.15.4.patch -> nix-etag.1.1x.patch

yorickvP committed:
nginx: use actual file path to calculate etag by nix hash

yorickvP commented:
updated: renamed the patch file in separate commits. Fixed API compatibility. Fixed memleaks.
@aszlig @yegortimoshenko 
https://github.com/NixOS/nixpkgs/pull/66532/commits/261f85d90b7d51022cfc9302b142e993f7c0465a is the relevant diff.

7c6f434c commented:
I wonder if it makes sense to allow NULL path and provide the old external API as a wrapper.

yorickvP commented:
@7c6f434c that's what I'm doing now, at least

7c6f434c commented on pkgs/servers/http/nginx/nix-etag-1.x.patch:
```diff
@@ -0,0 +1,167 @@
++
++        if (etag->value.data == NULL) {
++            etag->hash = 0;
++            if (real != NULL) ngx_free(real);
```
In other places you don't do a `NULL` check (and apparently at least `glibc` version of `free` is OK with `free(NULL)`) — why is this place different?

yorickvP replied:
I do a NULL check everywhere it could be null, which isn't in the other branch of the if

aszlig replied:
`ngx_free` is just an alias for `free` of the corresponding C library, from http://www.open-std.org/JTC1/SC22/wg14/www/docs/n1124.pdf section 7.20.3.2:

> If ptr is a null pointer, no action occurs.

So we don't need to do explicit checks like this.

7c6f434c commented on pkgs/servers/http/nginx/nix-etag-1.x.patch:
```diff
@@ -0,0 +1,167 @@
+ 
+     etag = ngx_list_push(&r->headers_out.headers);
+     if (etag == NULL) {
++        ngx_free(real);
```
If you do `NULL` checks before `free`, this seems reachable without setting `real` when `path` is `NULL`

domenkozar commented:
@yorickvP do you plan to address feedback?

cransom commented:
I'd like to see this get more attention/merged. Without it, my nginx config was unable to serve user directories (eg `example.com/~user/foo`) which caused some good head scratching for a while until I found the PR.

yorickvP commented:
Yes! I'll try to get to this this week.

Op wo 4 dec. 2019 om 23:12 schreef Casey Ransom <notifications@github.com>:

> I'd like to see this get more attention/merged. Without it, my nginx
> config was unable to serve user directories (eg example.com/~user/foo)
> which caused some good head scratching for a while until I found the PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/66532?email_source=notifications&email_token=AAE57JFND2L6QNZRSJ7LVQTQW7JFRA5CNFSM4ILCCUGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEF5RPMA#issuecomment-561715120>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAE57JFQ6S4WBCILUEWZYQTQW7JFRANCNFSM4ILCCUGA>
> .
>

jglukasik commented:
+1 on this getting more attention, this breaks my dynamic subdomain setup. I've been able to workaround this by overriding the `patches` attribute in the nginx package used by the service. Sharing my relevant `configuration.nix` lines below in case anyone else is having similar troubles:

```
services = {
  nginx = {
    enable = true;

    # don't apply patches, they break my subdomain setup
    # https://github.com/NixOS/nixpkgs/pull/66532
    package = (pkgs.nginx.overrideAttrs (oldAttrs: {
      name = "nginx-temp-no-patches";
      patches = [];
    }));

    virtualHosts = {
      "subs.ex.example.com" = {
        serverName = "~^(?<sub>.+)\\.ex\\.example\\.com$";
        root = "/var/www/ex.example.com/$sub";
      };
    };
  };
};

```

aszlig commented:
@jglukasik: See also #80671.
